### PR TITLE
[Maintenance] Upgrades TypeScript to 4.1, fixes simple errors

### PR DESCRIPTION
**Pull request by tmarkley**
### Description
* Reduces number of typescript errors from 517 across 260 files to 216 across 115 files (~58% reduction).
* The remaining errors are documented in #1660.
* TypeScript 4.1 introduces a set of [breaking changes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#breaking-changes). The main changes that impact Dashboards is that `resolve`'s parameters are no longer optional in `Promise`s, and that potentially undefined indexes must use the `!` non-null assertion operator.
 
### Issues Resolved
Incremental progress towards #1660.
 
### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
  - [x] `yarn test:jest_integration`
  - [x] `yarn test:ftr`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by kristenTian — DISMISSED** [review]

**Comment by ashwin-pc** [comment]
@tmarkley thanks for this! This was much overdue. I tried running the same tsc command that you did but saw only 15 errors and all of them were from the node modules folder. Do you mind sharing the steps to repro the errors?

**Comment by tmarkley** [comment]
> @tmarkley thanks for this! This was much overdue. I tried running the same tsc command that you did but saw only 15 errors and all of them were from the node modules folder. Do you mind sharing the steps to repro the errors?

Thanks for taking a look Ashwin! For sure. Cleaning and bootstrapping in `main` before running tsc works for me:

```
git checkout main
yarn osd clean
yarn osd bootstrap
tsc --noEmit
```

File: src/plugins/vis_type_table_new/public/components/table_vis_grid_columns.tsx [hunk 1]
```diff
@@ -18,26 +17,26 @@ export const getDataGridColumns = (
...
+  //             rows,
+  //           },
+  //           row: rowIndex,
+  //           column: rawColumnIndex,
+  //         },
+  //       ],
+  //       negate,
+  //     },
+  //   });
+  // };
```
Comment (ashwin-pc) [inline src/plugins/vis_type_table_new/public/components/table_vis_grid_columns.tsx:39]: Can we skip this too? @ananzh is in the process of moving this to replace the existing table Vis and will likely be updating this function.

**Review by ashwin-pc — COMMENTED** [review]
Thanks for taking the time for this massive PR @tmarkley. Almost all my concerns boil down to 3 things:
1. Do we need to use `error: any`? cant we instead use `e: unknown`? Im not a stickler about this but in many instance we rely on the error having a message prop that it might not sometimes have, which is why typescript went with the approach of requiring a strict type check for it. I'd love for the other maintainers to weigh in on this too. 
2. Skipping the Visbuilder type changes that are typed as `any`. Almost all of them an be strictly typed and can be tracked in the issue you have created to do just that
3. non null assertions. These too in many cases can be easily skipped. But i'd love to know your thoughts on them.

Overall, this is a massive step in the right direction, so thank you!

Also, I tried your exact steps to reproduce the type errors and got only 14. But that is most likely some issue in my `tsc` config

**Comment by tmarkley** [comment]
@ashwin-pc thanks for the review! Just getting to this after my extended holiday break 😄

File: src/plugins/vis_type_table_new/public/components/table_vis_grid_columns.tsx (hunk 1 above)
Comment (tmarkley) [reply src/plugins/vis_type_table_new/public/components/table_vis_grid_columns.tsx]: Yes, no longer relevant.

File: src/plugins/vis_builder/public/application/utils/breadcrumbs.ts [hunk 2]
```diff
@@ -10,7 +10,7 @@ const defaultEditText = i18n.translate('visBuilder.editor.defaultEditBreadcrumbT
   defaultMessage: 'Edit',
 });
 
-export function getVisualizeLandingBreadcrumbs(navigateToApp) {
+export function getVisualizeLandingBreadcrumbs(navigateToApp: any) {
```
[the remark this answers is not in the document]
Comment (tmarkley) [reply src/plugins/vis_builder/public/application/utils/breadcrumbs.ts:13]: Yes, I'll remove all of the changes in vis_builder! And yes, it is explicitly set to false: https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/vis_builder/tsconfig.json#L54

**Comment by joshuarrrr** [comment]
@tmarkley Is this ready for re-review, or are you still addressing some of @ashwin-pc's comments?

**Comment by joshuarrrr** [comment]
@tmarkley I'm placing this in draft status for now - feel free to move it out of draft when you're able to pick it back up.

**Comment by dblock** [comment]
Closing as stalled, please finish and reopen as you see fit. Thanks!

[Catch All Triage, attendees [1](https://github.com/dblock), [2](https://github.com/reta), [3](https://github.com/lukas-vlcek/), [4](https://github.com/rafaelma), [5](https://github.com/krisfreedain), [6](https://github.com/gargrajat), [7](https://github.com/andrross)]

**Pull request closed without merging**